lib/ain-ocean/src/api/pool_pair/service.rs

```diff
@@ -0,0 +1,360 @@
+
+    let mut a_token_rate = dec!(1);
+    let mut b_token_rate = dec!(1);
+
+    if p.id_token_a != usdt_id {
+        let BestSwapPathResponse {
+            estimated_return, ..
+        } = get_best_path(ctx, &p.id_token_a, &usdt_id).await?;
+        a_token_rate = Decimal::from_str(estimated_return.as_str())?;
+    }
+
+    if p.id_token_a != usdt_id {
+        let BestSwapPathResponse {
+            estimated_return, ..
+        } = get_best_path(ctx, &p.id_token_b, &usdt_id).await?;
+        b_token_rate = Decimal::from_str(estimated_return.as_str())?;
+    }
+
+    let reserve_a = Decimal::from_f64(p.reserve_a).unwrap_or_default();
+    let reserve_b = Decimal::from_f64(p.reserve_b).unwrap_or_default();
+
+    let a = a_token_rate
+        .checked_mul(reserve_a)
+        .ok_or_else(|| Error::OverflowError)?;
+
+    let b = b_token_rate
+        .checked_mul(reserve_b)
+        .ok_or_else(|| Error::OverflowError)?;
+
+    let res = a.checked_add(b).ok_or_else(|| Error::OverflowError)?;
+
+    Ok(res)
+}
+
+pub async fn get_total_liquidity_usd(ctx: &Arc<AppContext>, p: &PoolPairInfo) -> Result<Decimal> {
+    let parts = p.symbol.split('-').collect::<Vec<&str>>();
+    let [a, b] = <[&str; 2]>::try_from(parts).ok().unwrap();
+
+    let reserve_a = Decimal::from_f64(p.reserve_a).unwrap_or_default();
+    let reserve_b = Decimal::from_f64(p.reserve_b).unwrap_or_default();
+
+    if ["DUSD", "USDT", "USDC"].contains(&a) {
+        return reserve_a
+            .checked_mul(dec!(2))
+            .ok_or_else(|| Error::OverflowError);
+    };
+
+    if ["DUSD", "USDT", "USDC"].contains(&b) {
+        return reserve_b
+            .checked_mul(dec!(2))
+            .ok_or_else(|| Error::OverflowError);
+    };
+
+    let usdt_per_dfi = get_usd_per_dfi(ctx).await?;
+    if usdt_per_dfi.is_zero() {
+        return Ok(usdt_per_dfi);
+    };
+
+    if a == "DFI" {
+        return reserve_a
+            .checked_mul(dec!(2))
+            .ok_or_else(|| Error::OverflowError)?
+            .checked_mul(usdt_per_dfi)
+            .ok_or_else(|| Error::OverflowError);
+    };
+
+    if b == "DFI" {
+        return reserve_b
+            .checked_mul(dec!(2))
+            .ok_or_else(|| Error::OverflowError)?
+            .checked_mul(usdt_per_dfi)
+            .ok_or_else(|| Error::OverflowError);
+    };
+
+    let res = get_total_liquidity_usd_by_best_path(ctx, p).await?;
+    Ok(res)
+}
+
+async fn get_yearly_custom_reward_usd(ctx: &Arc<AppContext>, p: &PoolPairInfo) -> Result<Decimal> {
+    if p.custom_rewards.is_none() {
+        return Ok(dec!(0));
+    };
+
+    let dfi_price_usdt = get_usd_per_dfi(ctx).await?;
+    if dfi_price_usdt.is_zero() {
+        return Ok(dfi_price_usdt);
+    };
+
+    let rewards = if let Some(custom_rewards) = &p.custom_rewards {
+        let total = custom_rewards
+            .iter()
+            .fold(dec!(0), |accumulate: Decimal, account: &String| {
```

Comment:
Should use `try_fold` and return the Result instead of unwrapping here

Reply:
haven't got idea to `impl try_fold for Decimal` yet 😓 
replace `unwrap` by `expect` for better error msg temporary... 
https://github.com/DeFiCh/ain/pull/2838/commits/e53119bf45064ce3e4dff0c4e0e4ae4ee27d13cf